core/core.go

```diff
@@ -222,62 +258,235 @@ func NewTextile(ctx context.Context, conf Config) (*Textile, error) {
+		method, _ := grpc.Method(ctx)
+		var isUser bool
+		var owner crypto.PubKey
+		if org, ok := c.OrgFromContext(ctx); ok {
+			owner = org.Key
+		} else if dev, ok := c.DevFromContext(ctx); ok {
+			owner = dev.Key
+		} else if k, ok := common.APIKeyFromContext(ctx); ok {
+			key, err := t.co.Keys.Get(ctx, k)
+			if err != nil || !key.Valid {
+				return nil, status.Error(codes.NotFound, "API key not found or is invalid")
+			}
+			if !common.ValidateAPISigContext(ctx, key.Secret) {
+				return nil, status.Error(codes.PermissionDenied, "Bad API key signature")
+			}
+
+			// Pull user key out of the thread token if present
+			token, ok := thread.TokenFromContext(ctx)
+			if ok {
+				var claims jwt.StandardClaims
+				if _, _, err = new(jwt.Parser).ParseUnverified(string(token), &claims); err != nil {
+					return nil, status.Error(codes.PermissionDenied, "Bad authorization")
+				}
+				ukey := &thread.Libp2pPubKey{}
+				if err = ukey.UnmarshalString(claims.Subject); err != nil {
+					return nil, err
+				}
+				owner = ukey.PubKey
+				isUser = true
+			}
+			if getService(method) == "users.pb.API" {
+				if owner == nil {
+					return nil, status.Error(codes.Unauthenticated, "Token required")
+				}
+				user, err := t.co.Users.Get(ctx, owner)
+				if err != nil {
+					return nil, status.Error(codes.NotFound, "User not found")
+				}
+				ctx = c.NewUserContext(ctx, user)
+			}
+		}
+
+		// Let the request pass through
+		res, err := handler(ctx, req)
+		if err != nil {
+			return res, err
+		}
+
+		var createID, deleteID thread.ID
+		switch method {
+		case "/threads.pb.API/NewDB":
+			createID, err = thread.Cast(req.(*dbpb.NewDBRequest).DbID)
+			if err != nil {
+				return res, err
+			}
+		case "/threads.pb.API/NewDBFromAddr":
+			addr, err := ma.NewMultiaddrBytes(req.(*dbpb.NewDBFromAddrRequest).Addr)
+			if err != nil {
+				return res, err
+			}
+			createID, err = thread.FromAddr(addr)
+			if err != nil {
+				return res, err
+			}
+		case "/threads.pb.API/DeleteDB":
+			deleteID, err = thread.Cast(req.(*dbpb.DeleteDBRequest).DbID)
+			if err != nil {
+				return res, err
+			}
+		case "/threads.net.pb.API/CreateThread":
+			createID, err = thread.Cast(req.(*netpb.CreateThreadRequest).ThreadID)
+			if err != nil {
+				return res, err
+			}
+		case "/threads.net.pb.API/AddThread":
+			addr, err := ma.NewMultiaddrBytes(req.(*netpb.AddThreadRequest).Addr)
+			if err != nil {
+				return res, err
+			}
+			createID, err = thread.FromAddr(addr)
+			if err != nil {
+				return res, err
+			}
+		case "/threads.net.pb.API/DeleteThread":
+			deleteID, err = thread.Cast(req.(*netpb.DeleteThreadRequest).ThreadID)
+			if err != nil {
+				return res, err
+			}
+		default:
+			return res, nil
+		}
 
-	newCtx := context.WithValue(ctx, clientReqKey("user"), user)
-	newCtx = context.WithValue(newCtx, clientReqKey("project"), proj)
-	return newCtx, nil
+		if createID.Defined() {
```

Comment:
We've gotten this far, the caller either created or deleted a thread. Use the local threads table to keep track of what happened.